### PR TITLE
Require matching hashlen in RSA functions

### PR DESCRIPTION
**Pull request by mpg**
## Description

Functions in the RSA module that accept a hashlen parameter now require it to match the output size of the hash algorithm used, except when signing raw data.

This was already required in the PK module but not enforced there until #4561

~For now, this is just a documentation change. If we agree about this change, I'll have a go at implementing a corresponding check in the code, and see how much adaptations the tests need.~

Based on #4665 in order to avoid conflicts. Continues in https://github.com/ARMmbed/mbedtls/pull/4707 with the implementation.

## Status
**IN DEVELOPMENT**

## Requires Backporting
NO - API Change

## Migrations
YES - stronger requirements on the caller of some APIs

### CODE REVIEW

**Review by yanesca — APPROVED** [review]
Looks good to me.

File: docs/3.0-migration-guide.d/require-matching-hashlen-rsa.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+RSA signature functions now require `hashlen` to match the expected value
+-------------------------------------------------------------------------
+
+This only affects users of the low-level RSA API; users of the high-level PK
+API or of the PSA Crypto API are not affected.
+
+All the functions in the RSA module that accept a `hashlen` parameter used to
+ignore it unless the `md_alg` parameter was `MBEDTLS_MD_NONE`, indicating raw
+data was signed. They now require this parameter's value to be equal to the
+output size of the hash algorithm used when signing a hash. (The requirements
```
Comment (gilles-peskine-arm) [inline docs/3.0-migration-guide.d/require-matching-hashlen-rsa.md:10]: I think the salient point is that `hashlen` is now the size that is read from the buffer.

**Review by gilles-peskine-arm — COMMENTED** [review]
Looks good to me except for some non-blockers in the documentation. But having dug a little, I think it's incomplete: PK signature functions shouldn't accept `hash_len==0` anymore.

**Comment by gilles-peskine-arm** [comment]
> If we agree about this change, I'll have a go at implementing a corresponding check in the code, and see how much adaptations the tests need.

Done in https://github.com/ARMmbed/mbedtls/pull/4707. There wasn't much to change.

**Comment by mpg** [comment]
Superseded by #4707

**Comment by gilles-peskine-arm** [comment]
Closing in favor of #4707 since that seems to be in good enough shape to be merged by release time.

**Pull request closed without merging**